File: day-15/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

typedef struct {
    char *label;
    unsigned short box_;
    char operand;
    int focal_length;
} Label;

typedef struct {
    Label *labels;
    size_t size;
    size_t capacity;
} LabelArray;

typedef struct {
    unsigned short box_;
    LabelArray label_array;
} HashMapEntry;

typedef struct {
    HashMapEntry *entries;
    size_t size;
    size_t capacity;
} HashMap;

unsigned short hash_char(char c, unsigned short hash) {
    hash += (unsigned short)c;
    hash = (unsigned short)((hash * 17) % 256);
    return hash;
}

unsigned short hash_string(const char *s, unsigned short hash) {
    while (*s) {
        hash = hash_char(*s++, hash);
    }
    return hash;
}

void init_label_array(LabelArray *arr) {
    arr->size = 0;
    arr->capacity = 4;
    arr->labels = malloc(arr->capacity * sizeof(Label));
}

void append_label_array(LabelArray *arr, Label label) {
    if (arr->size >= arr->capacity) {
        arr->capacity *= 2;
        arr->labels = realloc(arr->labels, arr->capacity * sizeof(Label));
    }
    arr->labels[arr->size++] = label;
}

void init_hash_map(HashMap *map) {
    map->size = 0;
    map->capacity = 4;
    map->entries = malloc(map->capacity * sizeof(HashMapEntry));
}

void insert_hash_map(HashMap *map, unsigned short box_, Label label) {
    for (size_t i = 0; i < map->size; i++) {
        if (map->entries[i].box_ == box_) {
            LabelArray *labels = &map->entries[i].label_array;
            for (size_t j = 0; j < labels->size; j++) {
                if (strcmp(labels->labels[j].label, label.label) == 0) {
                    labels->labels[j].focal_length = label.focal_length;
                    return;
                }
            }
            append_label_array(labels, label);
            return;
        }
    }
    if (map->size >= map->capacity) {
        map->capacity *= 2;
        map->entries = realloc(map->entries, map->capacity * sizeof(HashMapEntry));
    }
    map->entries[map->size].box_ = box_;
    init_label_array(&map->entries[map->size].label_array);
    append_label_array(&map->entries[map->size].label_array, label);
    map->size++;
}

void remove_label(HashMap *map, unsigned short box_, const char *label) {
    for (size_t i = 0; i < map->size; i++) {
        if (map->entries[i].box_ == box_) {
            LabelArray *labels = &map->entries[i].label_array;
            for (size_t j = 0; j < labels->size; j++) {
                if (strcmp(labels->labels[j].label, label) == 0) {
                    for (size_t k = j; k < labels->size - 1; k++) {
                        labels->labels[k] = labels->labels[k + 1];
                    }
                    labels->size--;
                    return;
                }
            }
        }
    }
}

Label parse_label(const char *complete_label) {
    Label label;
    label.label = malloc(strlen(complete_label) + 1);
    strcpy(label.label, complete_label);
    
    char *operand_pos = strchr(label.label, '=');
    if (operand_pos) {
        label.operand = '=';
    } else {
        operand_pos = strchr(label.label, '-');
        if (operand_pos) {
            label.operand = '-';
        }
    }
    
    if (operand_pos) {
        *operand_pos = '\0';
        label.focal_length = atoi(operand_pos + 1);
    } else {
        label.focal_length = 0;
    }
    
    label.box_ = hash_string(label.label, 0);
    return label;
}

int calc_focusing_power(const char *input) {
    char *input_copy = strdup(input);
    char *token = strtok(input_copy, ",");
    HashMap hash_map;
    init_hash_map(&hash_map);
    
    while (token) {
        Label label = parse_label(token);
        if (label.operand == '-') {
            remove_label(&hash_map, label.box_, label.label);
        } else if (label.operand == '=') {
            insert_hash_map(&hash_map, label.box_, label);
        } else {
            fprintf(stderr, "Invalid operand\n");
            exit(EXIT_FAILURE);
        }
        token = strtok(NULL, ",");
    }
    
    int focusing_power = 0;
    for (size_t i = 0; i < hash_map.size; i++) {
        unsigned short box_ = hash_map.entries[i].box_;
        LabelArray *labels = &hash_map.entries[i].label_array;
        for (size_t j = 0; j < labels->size; j++) {
            int box_value = (box_ + 1);
            int index_value = (j + 1);
            focusing_power += box_value * index_value * labels->labels[j].focal_length;
        }
    }
    
    free(input_copy);
    return focusing_power;
}

int main() {
    FILE *file = fopen("../input.txt", "r");
    if (!file) {
        fprintf(stderr, "Unable to read file\n");
        return EXIT_FAILURE;
    }

    fseek(file, 0, SEEK_END);
    long fsize = ftell(file);
    fseek(file, 0, SEEK_SET);

    char *input = malloc(fsize + 1);
    fread(input, 1, fsize, file);
    fclose(file);
    input[fsize] = '\0';

    char *newline = strrchr(input, '\n');
    if (newline) {
        *newline = '\0';
    }

    clock_t start = clock();
    int response = calc_focusing_power(input);
    clock_t end = clock();
    
    printf("Focusing power: %d\n", response);
    printf("Time taken: %ldms\n", (end - start) * 1000 / CLOCKS_PER_SEC);

    free(input);
    return 0;
}

File: day-15/rust/src/main.rs
use std::collections::HashMap;
use std::fs;

#[derive(Clone, Debug)]
struct Label {
    label: String,
    box_: u16,
    operand: char,
    focal_length: i32,
}

fn hash_char(c: char, mut hash: u16) -> u16 {
    hash += c as u16;
    hash = hash.wrapping_mul(17);
    hash %= 256;
    hash
}

fn hash_string(s: &str, mut hash: u16) -> u16 {
    for c in s.chars() {
        hash = hash_char(c, hash);
    }
    hash
}

fn assign_lens(label: Label, hash_map: &mut HashMap<u16, Vec<Label>>) {
    match label.operand {
        '-' => {
            if let Some(labels) = hash_map.get_mut(&label.box_) {
                labels.retain(|l| l.label != label.label);
            }
        }
        '=' => {
            let labels = hash_map.entry(label.box_).or_insert_with(Vec::new);
            if let Some(existing_label) = labels.iter_mut().find(|l| l.label == label.label) {
                existing_label.focal_length = label.focal_length;
            } else {
                labels.push(label);
            }
        }
        _ => panic!("Invalid operand"),
    }
}

fn parse_label(complete_label: &str) -> Label {
    let operand = if complete_label.contains('=') {
        '='
    } else {
        '-'
    };
    let parts: Vec<&str> = complete_label.split(operand).collect();
    let label = parts[0].to_string();
    let focal_length = parts.get(1).and_then(|s| s.parse().ok()).unwrap_or(0);
    let box_ = hash_string(&label, 0);
    Label {
        label,
        box_,
        operand,
        focal_length,
    }
}

fn calc_focusing_power(input: &str) -> i32 {
    let labels: Vec<&str> = input.split(',').collect();
    let mut hash_map: HashMap<u16, Vec<Label>> = HashMap::new();

    for label in labels {
        let parsed_label = parse_label(label);
        assign_lens(parsed_label, &mut hash_map);
    }

    let mut focusing_power = 0;
    for (box_, labels) in hash_map {
        if labels.is_empty() {
            continue;
        }
        for (index, label) in labels.iter().enumerate() {
            let box_value = (box_ as i32) + 1;
            let index_value = (index as i32) + 1;
            focusing_power += box_value * index_value * label.focal_length;
        }
    }

    focusing_power
}

fn main() {
    let input = fs::read_to_string("/Users/carlosgonzalez/personal/advent-of-code-2023/day-15/input.txt").expect("Unable to read file");
    let input = input.trim();

    let start = std::time::Instant::now();
    let response = calc_focusing_power(&input);
    println!("Focusing power: {}", response);
    println!("Time taken: {}ms", start.elapsed().as_millis());
}
